shared/src/ledger/eth_bridge/storage/bridge_pool.rs

```diff
@@ -44,3 +58,847 @@ pub fn is_bridge_pool_key(key: &Key) -> bool {
 pub fn is_protected_storage(key: &Key) -> bool {
     is_bridge_pool_key(key) && *key != get_pending_key()
 }
+
+/// A simple Merkle tree for the Ethereum bridge pool
+#[derive(
+    Debug, Default, Clone, BorshSerialize, BorshDeserialize, BorshSchema,
+)]
+pub struct BridgePoolTree {
+    /// Root of the tree
+    root: KeccakHash,
+    /// The underlying storage
+    store: BTreeSet<KeccakHash>,
+}
+
+impl BridgePoolTree {
+    /// Create a new merkle tree for the Ethereum bridge pool
+    pub fn new(root: KeccakHash, store: BTreeSet<KeccakHash>) -> Self {
+        Self { root, store }
+    }
+
+    /// Parse the key to ensure it is of the correct type.
+    ///
+    /// If it is, it can be converted to a hash.
+    /// Checks if the hash is in the tree.
+    pub fn contains_key(&self, key: &Key) -> Result<bool, Error> {
+        Ok(self.store.contains(&Self::parse_key(key)?))
+    }
+
+    /// Update the tree with a new value.
+    ///
+    /// Returns the new root if successful. Will
+    /// return an error if the key is malformed.
+    pub fn update_key(&mut self, key: &Key) -> Result<Hash, Error> {
+        let hash = Self::parse_key(key)?;
+        _ = self.store.insert(hash);
+        self.root = self.compute_root();
+        Ok(self.root())
+    }
+
+    /// Delete a key from storage and update the root
+    pub fn delete_key(&mut self, key: &Key) -> Result<(), Error> {
+        let hash = Self::parse_key(key)?;
+        _ = self.store.remove(&hash);
+        self.root = self.compute_root();
+        Ok(())
+    }
```

Comment:
We could do this logging here, or in the VP, or when we apply the txs in the ledger. What do you think is the best place?

Reply:
If we want to log whether the key was newly inserted/deleted, that information is only in this function so we'd have to log here, or alternatively we could change these fns to return `Result<bool>` where the returned bool is whether the key was newly inserted/deleted (though then the callers of `update_key` would have to fetch the root hash themselves after calling `update_key`). 

Re. where to log, logging it in the native VP would be more practical I think as logging inside wasms is less functional it seems. I guess the main thing is we should be able to see when running a node what is happening with the bridge pool, to help us with development and debugging. I'd be tempted even to log at INFO so that real validators/nodes should always have this logged as well, in case we need to debug something later on in a testnet, they could give us the logs

Reply:
If we log at this point, it could be confusing. Because this code is called when changing the write ahead log, which might then gets rejected by the VP and so this update/delete doesn't actually happen. I think I'll log at the VP level.